File: problem-solving/baekjoon/16165_걸그룹_마스터_준석이.py
# https://www.acmicpc.net/problem/16165
import sys

if __name__ == "__main__":
    n, m = list(map(int, sys.stdin.readline().split()))
    group_infos = {}
    member_infos = {}
    for i in range(n):
        group_name = sys.stdin.readline().rstrip()
        member_count = int(sys.stdin.readline())
        members = []
        for j in range(member_count):
            member_name = sys.stdin.readline().rstrip()
            members.append(member_name)
            member_infos[member_name] = group_name
        group_infos[group_name] = (group_name, members)
    for i in range(m):
        quiz = sys.stdin.readline().rstrip()
        quiz_type = int(sys.stdin.readline())
        if quiz_type == 0:
            for j in sorted(group_infos[quiz][1]):
                print(j)
        elif quiz_type == 1:
            print(member_infos[quiz])

File: problem-solving/baekjoon/16165_걸그룹_마스터_준석이.go
// https://www.acmicpc.net/problem/16165
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	var n, m int
	fmt.Fscanln(reader, &n, &m)
	groupInfos := map[string]groupInfo{}
	memberInfos := map[string]string{}
	for i := 0; i < n; i++ {
		var groupName string
		fmt.Fscanln(reader, &groupName)
		var memberCount int
		fmt.Fscanln(reader, &memberCount)
		info := groupInfo{groupName: groupName}
		for j := 0; j < memberCount; j++ {
			var memberName string
			fmt.Fscanln(reader, &memberName)
			info.members = append(info.members, memberName)
			memberInfos[memberName] = groupName
		}
		groupInfos[groupName] = info
	}
	for i := 0; i < m; i++ {
		var quiz string
		fmt.Fscanln(reader, &quiz)
		var quizType int
		fmt.Fscanln(reader, &quizType)
		if quizType == 0 {
			sort.Strings(groupInfos[quiz].members)
			for _, v := range groupInfos[quiz].members {
				fmt.Fprintln(writer, v)
			}
		} else if quizType == 1 {
			fmt.Fprintln(writer, memberInfos[quiz])
		}
	}
}

type groupInfo struct {
	groupName string
	members   []string
}
